clippy_config/src/de.rs

```diff
@@ -0,0 +1,634 @@
+        T::deserialize(dcx, value).map(|x| Spanned {
+            node: x,
+            span: dcx.make_sp(value.span()),
+        })
+    }
+}
+
+/// A type which can be deserialized from a toml value with a fallback to a default value.
+pub trait DeserializeOrDefault<T>: Sized {
+    /// Attempt to deserialize the value. Returns the default value and raises an error on failure.
+    fn deserialize_or_default(dcx: &DiagCtxt<'_>, value: Item<'_>, default: T) -> Self;
+}
+
+impl<T: Deserialize + FromDefault<T>> DeserializeOrDefault<T> for T {
+    fn deserialize_or_default(dcx: &DiagCtxt<'_>, value: Item<'_>, default: T) -> Self {
+        T::deserialize(dcx, value).unwrap_or_else(|| T::from_default(default))
+    }
+}
+
+impl<T: Deserialize + Default> DeserializeOrDefault<()> for T {
+    fn deserialize_or_default(dcx: &DiagCtxt<'_>, value: Item<'_>, (): ()) -> Self {
+        T::deserialize(dcx, value).unwrap_or_default()
+    }
+}
+
+impl<T: Deserialize> DeserializeOrDefault<()> for Option<T> {
+    fn deserialize_or_default(dcx: &DiagCtxt<'_>, value: Item<'_>, (): ()) -> Self {
+        T::deserialize(dcx, value)
+    }
+}
+
+pub fn deserialize_array<T, U, C>(dcx: &DiagCtxt<'_>, value: Item<'_>, default: &'static [U]) -> C
+where
+    T: Deserialize + FromDefault<U>,
+    U: Copy,
+    C: FromIterator<T> + Extend<T>,
+{
+    let default_iter = default.iter().map(|&x| T::from_default(x));
+    if let Some((_, values)) = value.as_array() {
+        let mut has_default = false;
+        let mut res: C = values
+            .filter(|x| {
+                if x.as_str().is_some_and(|x| x == "..") {
+                    if has_default {
+                        dcx.span_warn(value.span(), "duplicate `..` item");
+                    }
+                    has_default = true;
+                    false
+                } else {
+                    true
+                }
+            })
+            .filter_map(|x| T::deserialize(dcx, x))
+            .collect();
+        if has_default {
+            res.extend(default_iter);
+        }
+        res
+    } else {
+        dcx.span_err(value.span(), "expected an array");
+        default_iter.collect()
+    }
+}
+
+impl<T, U> DeserializeOrDefault<&'static [U]> for Vec<T>
+where
+    T: Deserialize + FromDefault<U>,
+    U: Copy,
+{
+    fn deserialize_or_default(dcx: &DiagCtxt<'_>, value: Item<'_>, default: &'static [U]) -> Self {
+        deserialize_array(dcx, value, default)
+    }
+}
+
+impl<T, U, S> DeserializeOrDefault<&'static [U]> for HashSet<T, S>
+where
+    T: Deserialize + FromDefault<U> + Eq + Hash,
+    U: Copy,
+    S: Default + BuildHasher,
+{
+    fn deserialize_or_default(dcx: &DiagCtxt<'_>, value: Item<'_>, default: &'static [U]) -> Self {
+        deserialize_array(dcx, value, default)
+    }
+}
+
+macro_rules! deserialize_table {
+    ($dcx:ident, $table:ident, $($name:ident($name_str:literal): $ty:ty,)+) => {
```

Comment:
NIT: I think the usage if this macro is not self-explanatory. I would suggest adding a doc comment and renaming it to `deserialize_table_row` since it only processes one row.

Maybe:

```suggestion
/// Can be used to deserialize a table row. Each column has a local
/// of type `Option<T>` that is filled, when the value is found.
///
/// Example usage:
/// ```
/// deserialize_table_row!(dcx, table,
///     path("path"): String,
///     reason("reason"): String,
/// );
/// ```
///
/// Example input:
/// ```toml
/// disallowed-methods = [
///     { path = "std::vec::Vec::leak", reason = "no leaking memory" },
///     # path   = Some("std::vec::Vec::leak")
///     # reason = Some("no leaking memory")
///
///     { path = "std::time::Instant::now" },
///     # path   = Some("std::vec::Vec::leak")
///     # reason = None
/// ]
/// ```
macro_rules! deserialize_table_row {
    ($dcx:ident, $table:ident, $($name:ident($name_str:literal): $ty:ty,)+) => {
```

Reply:
What's this about rows? It can deserialize any toml table, both inline and regular. The `TableLike` trait is implemented for both.

Reply:
Huh, I guess it depends on what you describe as a "table". I thought the whole `disallowed-methods` is a table, like this

| path | reason |
|---|---|
| `"std::vec::Vec::leak"` | `"std::vec::Vec::leak"` |
| `"std::time::Instant::now"` | |

But I'm guessing now that you called `{ path = "std::vec::Vec::leak", reason = "no leaking memory" }` a table?

Reply:
Ah, that's what you mean. The name is referring to a toml table.

Reply:
Okay, then we can keep it like this. Could you rename the commit and fix the last error? Then we can merge this :D